### ISSUE TITLE
msiTarFileExtract (or ibun) attempts to create directories on wrong server 

### ISSUE THREAD

**Issue by erwinvanwieringen** [irods/irods#5410]
### What did you try to do?
I try to extract a TAR file, created with msiTarFileCreate, that has two replicas in resources configured on different resource servers. I specify the resource where I want the extracted objects explicitly. 

### Expected behavior
Files get extracted and registered as dataobjects, on the resource i specified.

### Observed behavior (including steps to reproduce, if applicable)
When I use ibun, it failes with error:

    remote addresses: 131.224.35.42 ERROR: bunUtil: opr error for /demoZone/home/rods/dataset, status = -520013 status = -520013 UNIX_FILE_MKDIR_ERR, Permission denied

If using msiTarFileExtract, a similar error appears in the rodsLog.

### To reproduce:
Install two irods servers, one provider (*server1*) and one consumer (*server2*)
Create a resource on each of the servers:
on *server1*:

    mkdir -p /vault1/primary
    iadmin mkresc primaryResc unixfilesystem server1:/vault1/primary

on *server2*:

    mkdir -p /vault2/faraway
    iadmin mkresc farawayResc unixfilesystem server2:/vault2/faraway

now create a collection with a dataobject in it

    imkdir dataset
    echo TEST > testfile
    iput testfile /demoZone/home/rods/dataset/testfile

create an archive file of my dataset. I use msiTarFileCreate, because ibun prepends an extra directory name

    irule  'msiTarFileCreate("/demoZone/home/rods/dataset.tar", "/demoZone/home/rods/dataset", "primaryResc", "")' null null

now i replicate my tar file to the *farawayResc* and remove the original data object

    irepl -R farawayResc /demoZone/home/rods/dataset.tar
    irm -rf /demoZone/home/rods/dataset

now  try the extraction:

    ibun -x -R primaryResc /demoZone/home/rods/dataset.tar /demoZone/home/rods/dataset

and this fails with the error shown above. What is more: the rodsLog of *server2* will show:

    Feb 14 15:06:12 pid:45483 remote addresses: <ip-hidden> ERROR: [-]	/irods_git_repo/server/core/src/fileOpr.cpp:252:int mkFileDirR(rsComm_t *, size_t, const std::string &, const std::string &, int) :  status [UNIX_FILE_MKDIR_ERR]  errno [Permission denied] -- message [fileMkdir for [/vault1]]
	[-]	/irods_git_repo/server/drivers/src/fileDriver.cpp:290:irods::error fileMkdir(rsComm_t *, irods::first_class_object_ptr) :  status [UNIX_FILE_MKDIR_ERR]  errno [Permission denied] -- message [failed to call 'mkdir']
		[-]	/irods_git_repo/plugins/resources/unixfilesystem/libunixfilesystem.cpp:989:irods::error unix_file_mkdir(irods::plugin_context &) :  status [UNIX_FILE_MKDIR_ERR]  errno [Permission denied] -- message [Mkdir error for "/vault1", errno = "Permission denied", status = -520013.]
     
    Feb 14 15:06:12 pid:45483 remote addresses: <ip-hidden> ERROR: mkFileDirR failed in procCacheDir with status -520013
    Feb 14 15:06:12 pid:45483 remote addresses: <ip-hidden> ERROR: _rsStructFileExtract - failed in call to procCacheDir status = -520013

Apparently it tries to create the directory path associated with my *primaryResc* on *server2*, while it should be created on *server1*. The error comes from the fact that filesystem permissions do not allow the directory /vault1 to be created. If i create a directory /vault1 on *server2*, owned by my irods service account, it creates the directory tree on this *server2*, while it should be on *server1*

**Comment by trel** [comment]
@erwinvanwieringen  Can you confirm your icommands are configured to connect to server2 in the case above?

Does it work as expected if you run the `ibun -x` while connected to server1?

If so, this is a 'lack of redirection' bug.

If not, we'll need to dig a bit more.

**Comment by erwinvanwieringen** [comment]
Apparently I missed a part of the documentation that says that the structfile does have to have a replica on the extraction resource. So I tried replicating my dataset.tar back to resource primaryResc. However, when I try the ibun -x -R command then, I still get the error shown above, when extracting to my `primaryResc`
```
client:> ils -L
/demoZone/home/rods:
  rods          1 primaryResc        10240 2021-12-06.16:11 & dataset.tar
    sha2:JozLhnm4vXtsQfNafs1nL7j8KqE7JkVe5tNQCB6/BK8=    generic    /vault2/faraway/home/rods/dataset.tar
  rods          2 farawayResc        10240 2021-12-06.16:29 & dataset.tar
    sha2:JozLhnm4vXtsQfNafs1nL7j8KqE7JkVe5tNQCB6/BK8=    generic    /vault1/primary/home/rods/dataset.tar
client:> ibun -x -R farawayResc /demoZone/home/rods/dataset.tar /demoZone/home/rods/dataset
client:> ils dataset
/demoZone/home/rods/dataset:
  testfile
client:> irm -rf dataset
client:> ibun -x -R primaryResc /demoZone/home/rods/dataset.tar /demoZone/home/rods/dataset
remote addresses: <server1_ip> ERROR: bunUtil: opr error for /demoZone/home/rods/dataset, status = -520013 status = -520013 UNIX_FILE_MKDIR_ERR, Permission denied
client:>
```
rodsLog of server1:
```
Dec  6 16:23:03 pid:49083 remote addresses: 127.0.0.1, <client_ip>, <server2_ip> ERROR: unbunPhyBunFile: rsStructFileExtract err for /demoZone/home/rods/dataset.tar. status = -520013
Dec  6 16:23:03 pid:49083 remote addresses: 127.0.0.1, <client_ip>, <server2_ip> ERROR: rsStructFileExtAndReg:unbunPhyBunFile err for /vault1/primary/home/rods/dataset.tar to dir /vault1/primary/home/rods/dataset.tar.dir.stat=-520013
```
rodsLog of server2:
```
Dec  6 16:23:03 pid:43645 remote addresses: <server1_ip> ERROR: [-]     /irods_git_repo/server/core/src/fileOpr.cpp:252:int mkFileDirR(rsComm_t *, size_t, const std::string &, const std::string &, int) :  status [UNIX_FILE_MKDIR_ERR]  errno [Permission denied] -- message [fileMkdir for [/vault1]]
        [-]     /irods_git_repo/server/drivers/src/fileDriver.cpp:290:irods::error fileMkdir(rsComm_t *, irods::first_class_object_ptr) :  status [UNIX_FILE_MKDIR_ERR]  errno [Permission denied] -- message [failed to call 'mkdir']
                [-]     /irods_git_repo/plugins/resources/unixfilesystem/libunixfilesystem.cpp:989:irods::error unix_file_mkdir(irods::plugin_context &) :  status [UNIX_FILE_MKDIR_ERR]  errno [Permission denied] -- message [Mkdir error for "/vault1", errno = "Permission denied", status = -520013.]
```
The structfile has a copy on both resources, but I can only extract it on `farawayResc`. If I extract on `primaryResc`, an attempt is made to create a directory associated with the vault path of `primaryResc` on the server associated with `farawayResc`.

If I now rename `primaryResc` to something like `dataResc`, the situation  is reversed:
```
client:> iadmin modresc primaryResc name dataResc
If you modify a resource name, you and other users will need to
change your irods_environment.json files to use it, you may need to update
server_config.json and, if rules use the resource name, you'll need to
update the core rules (core.re).  This command will update various
tables with the new name.
Do you really want to modify the resource name? (enter y or yes to do so):y
OK, performing the resource rename
client:> ils -L
/demoZone/home/rods:
  rods          1 dataResc        10240 2021-12-06.16:11 & dataset.tar
    sha2:JozLhnm4vXtsQfNafs1nL7j8KqE7JkVe5tNQCB6/BK8=    generic    /vault2/faraway/home/rods/dataset.tar
  rods          2 farawayResc        10240 2021-12-06.16:29 & dataset.tar
    sha2:JozLhnm4vXtsQfNafs1nL7j8KqE7JkVe5tNQCB6/BK8=    generic    /vault1/primary/home/rods/dataset.tar
  C- /demoZone/home/rods/dataset
client:> ibun -x -R dataResc /demoZone/home/rods/dataset.tar /demoZone/home/rods/dataset
client:> ils dataset
/demoZone/home/rods/dataset:
  testfile
client:> irm -rf dataset
client:> ibun -x -R farawayResc /demoZone/home/rods/dataset.tar /demoZone/home/rods/dataset
remote addresses: <server1_ip> ERROR: bunUtil: opr error for /demoZone/home/rods/dataset, status = -520013 status = -520013 UNIX_FILE_MKDIR_ERR, Permission denied
```
It seems like the extraction process is always run on the server associated with the resource that is alphabetically the first in the list of resources where structFile has a copy.

**Comment by peterverraedt** [comment]
After fixing #6070 I see strange behaviour in <https://github.com/irods/irods/blob/master/server/core/src/irods_resource_redirect.cpp#L478-L492>. I have
* irods.example.com: irods provider without resources
* resc0.irods.example.com
* resc1.irods.example.com
and a replicated resource on resc0/resc1. When doing ibun -x on a tar file stored on the replicated resource, but on the irods provider, with the modified code below, the error log shows two lines, first "resc0.irods.example.com" and then "irods.example.com".

```
hostName_t* tmp_host  = last_resc_host->hostName;
while ( tmp_host ) {
    std::string name( tmp_host->name );
    if ( name.find( host_name ) != std::string::npos ) { // Fix for #6070
    irods::log(LOG_ERROR, name); // This shows first resc0.irods.example.com and then irods.example.com
    if ( name.compare(host_name) == 0 ) {                
        match_flg = true;
        break;
    }

    tmp_host = tmp_host->next;

    // Uncomment to "fix" behaviour
    // break;
} // while tmp_host
```

I can add a break in the loop to fix the behaviour, but I'm not sure how the resource "next" should iterate (over the resource hierarchy? - in this case the second iteration of the loop is a passthru resource, and the third would be the replication).

**Comment by trel** [comment]
@korydraughn is this ... fixed?   Did we decide this had the same root cause as #6070?

**Comment by korydraughn** [comment]
I believe this is fixed by the #6070 fix, but I haven't spent any time verifying that.

I will look into this.

**Comment by korydraughn** [comment]
I've confirmed that this issue was resolved by the #6070 fix.

This covered the following *(with and without the fix)*:
- Using `ibun -x -R` while connected to server1
- Using `ibun -x -R` while connected to server2

The tests were done using 4.2.10 and tip of 4-2-stable.
